api_core/google/api_core/grpc_helpers.py

```diff
@@ -136,3 +138,178 @@ def create_channel(target, credentials=None, scopes=None, **kwargs):
+            'Method stub for "{}" has no responses.'.format(self._method))
+
+
+def _simplify_method_name(method):
+    """Simplifies a gRPC method name.
+
+    When gRPC invokes the channel to create a callable, it gives a full
+    method name like "/google.pubsub.v1.Publisher/CreateTopic". This
+    returns just the name of the method, in this case "CreateTopic".
+
+    Args:
+        method (str): The name of the method.
+
+    Returns:
+        str: The simplified name of the method.
+    """
+    return method.rsplit('/', 1).pop()
+
+
+class ChannelStub(grpc.Channel):
+    """A testing stub for the grpc.Channel interface.
+
+    This can be used to test any client that eventually uses a gRPC channel
+    to communicate. By passing in a channel stub, you can configure which
+    responses are returned and track which requests are made.
+
+    For example:
+
+    .. code-block:: python
+
+        channel_stub = grpc_helpers.ChannelStub()
+        channel_stub.GetFoo.response = foo_pb2.Foo(name='bar')
+
+        client = FooClient(channel_stub)
+
+        foo = client.get_foo(labels=['baz'])
+
+        assert foo.name == 'bar'
+        assert channel_stub.GetFoo.requests[0].labels = ['baz']
+
+    Each method on the stub can be accessed and configured on the channel.
+    Here's some examples of various configurations:
+
+    .. code-block:: python
+
+        # Return a basic response:
+
+        channel_stub.GetFoo.response = foo_pb2.Foo(name='bar')
+        assert client.get_foo().name == 'bar'
+
+        # Raise an exception:
+        channel_stub.GetFoo.response = NotFound('...')
+
+        with pytest.raises(NotFound):
+            client.get_foo()
+
+        # Use a sequence of responses:
+        channel_stub.GetFoo.responses = iter([
+            foo_pb2.Foo(name='bar'),
+            foo_pb2.Foo(name='baz')])
+
+        assert client.get_foo().name == 'bar'
+        assert client.get_foo().name == 'baz'
+
+        # Use a callable
+
+        def on_get_foo(request):
+            return foo_pb2.Foo(name='bar' + request.id)
+
+        channel_stub.GetFoo.response = on_get_foo
+
+        assert client.get_foo(id='123').name == 'bar123'
+    """
+
+    def __init__(self, responses=[]):
+        self._requests = []
+        self._method_stubs = {}
+
+    @property
+    def requests(self):
+        """Sequence[Tuple[str, protobuf.Message]]: Returns a list of all
+        requests made on this channel in order. The tuple is of method name,
+        request proto."""
+        return self._requests
+
+    def _stub_for_method(self, method):
+        method = _simplify_method_name(method)
+        if method not in self._method_stubs:
+            self._method_stubs[method] = _CallableStub(method, self)
+
+        return self._method_stubs[method]
+
+    def __getattr__(self, key):
+        # Ideally this would only return stubs for known methods, however,
+        # grpc doesn't actually create these methods until they're needed.
```

Comment:
Is that true?

Reply:
Seems not to be in [`proto`][1] or [`gapic`][2].

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/04308483f852d3cf551d2e33f5528bfb3e9650ac/datastore/google/cloud/datastore_v1/proto/datastore_pb2_grpc.py#L17-L57
[2]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/04308483f852d3cf551d2e33f5528bfb3e9650ac/datastore/google/cloud/datastore_v1/gapic/datastore_client.py#L115-L149

Reply:
Yeah I was doing things wrong before. Fixed.